File: LeetCode/Easy/Minimum Distance Between BST Nodes/Go/minimumDistanceBetweenBSTNodes.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val} }

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDiffInBST(root *TreeNode) int {
	sortedElements := []int{}
	inorder(&sortedElements, root)

	minDiff := sortedElements[1] - sortedElements[0]
	for i := 2; i < len(sortedElements); i++ {
		minDiff = min(minDiff, sortedElements[i]-sortedElements[i-1])
	}
	return minDiff
}

func inorder(sortedElements *[]int, root *TreeNode) {
	if root == nil {
		return
	}
	inorder(sortedElements, root.Left)
	*sortedElements = append(*sortedElements, root.Val)
	inorder(sortedElements, root.Right)
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func main() {
	root := New(4)
	root.Left = New(2)
	root.Right = New(6)
	root.Left.Left = New(1)
	root.Left.Right = New(3)
	fmt.Println(minDiffInBST(root))
}

File: LeetCode/Easy/Minimum Distance Between BST Nodes/Java/minimumDistanceBetweenBSTNodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> sortedElement;

    public int minDiffInBST(TreeNode root) {
        sortedElement = new ArrayList<>();
        inorder(root);

        int minDiff = sortedElement.get(1) - sortedElement.get(0);
        for (int i=2; i<sortedElement.size(); i++) 
            minDiff = Math.min(minDiff, sortedElement.get(i)-sortedElement.get(i-1));
        
        return minDiff;
    }

    public void inorder(TreeNode root) {
        if (root == null) return;
        inorder(root.left);
        sortedElement.add(root.val);
        inorder(root.right);
    }
}
